### PR TITLE
core: Fix memory leak

### PR DESCRIPTION
**Pull request by vonchenplus**
1. Fix KScopedAutoObject leak when SendSyncRequest
  **Because the KScopedSchedulerLock triggers a fiber switch. when a suspend been triggered, the current KScopedAutoObject reference count may not be released**
2. Fix SharedMemory leak

Known Issue:
1. svc WaitSynchronization objects may not be released correctly
**KSynchronizationObject::Wait will trigger a fiber switch, causing the object not to be released**
2. svc StartThread thread leak
**It is possible that the shutdown was performed without calling ExitThread**
3. HLE Service's event leak
**The event object is treated as a member parameter, when the class is released, the event object will be destructured and needs to be changed to event pointer object**

I tried to solve the three problems above, but I found that sometimes it would run crash, so I'll put it aside for now

### CODE REVIEW

**Comment by FernandoS27** [comment]
Let me know when it's ready for testing.

**Comment by vonchenplus** [comment]
> Let me know when it's ready for testing.

it's ready for testing

**Comment by FernandoS27** [comment]
Which source are you using for this ? It seems inaccurate. Best is to use the main source:
https://github.com/Atmosphere-NX/Atmosphere/tree/master/libraries/libmesosphere

we try to keep everything as close as possible since mesosphere is Hardware tested.

**Comment by vonchenplus** [comment]
My changes do not add any third-party libraries

File: src/core/hle/service/acc/async_context.cpp [hunk 1]
```diff
@@ -7,11 +7,10 @@
 #include "core/hle/service/acc/async_context.h"
 
 namespace Service::Account {
-IAsyncContext::IAsyncContext(Core::System& system_)
-    : ServiceFramework{system_, "IAsyncContext"}, compeletion_event{system_.Kernel()} {
+IAsyncContext::IAsyncContext(Core::System& system_) : ServiceFramework{system_, "IAsyncContext"} {
 
-    Kernel::KAutoObject::Create(std::addressof(compeletion_event));
-    compeletion_event.Initialize("IAsyncContext:CompletionEvent");
+    compeletion_event = Kernel::KEvent::Create(system_.Kernel());
```
Comment (bunnei) [inline src/core/hle/service/acc/async_context.cpp:12]: This and all others should use `KernelHelpers::ServiceContext::CreateEvent`

File: src/core/hle/kernel/k_process.cpp [hunk 2]
```diff
@@ -254,10 +255,28 @@ ResultCode KProcess::AddSharedMemory(KSharedMemory* shmem, [[maybe_unused]] VAdd
...
-    // Open a reference to the shared memory.
+        shemen_info->Initialize(shmem);
+        shared_memory_list.push_back(shemen_info);
+    }
+
+    // Open a reference to the shared memory and its info.
     shmem->Open();
+    shemen_info->Open();
+
+    // shared_memory_list.push_back(shmem);
```
Comment (bunnei) [inline src/core/hle/kernel/k_process.cpp:279]: Remove

File: src/core/hle/kernel/k_process.cpp [hunk 3]
```diff
@@ -412,6 +444,24 @@ void KProcess::Finalize() {
     // Finalize the handle table and close any open handles.
     handle_table.Finalize();
 
+    /* Free all shared memory infos. */
```
Comment (bunnei) [inline src/core/hle/kernel/k_process.cpp:447]: Comment style is incorrect (see one ones above)

File: src/core/hle/service/acc/async_context.cpp [hunk 4]
```diff
@@ -25,12 +24,17 @@ IAsyncContext::IAsyncContext(Core::System& system_)
     RegisterHandlers(functions);
 }
 
+IAsyncContext::~IAsyncContext() {
```
Comment (bunnei) [inline src/core/hle/service/acc/async_context.cpp:27]: Here and elsewhere below should use `KernelHelpers::ServiceContext::CloseEvent`

**Review by bunnei — CHANGES_REQUESTED** [review]
Looks mostly fine -- thank you for tracking down these leaks!

The main comment I have is to use KernelHelpers::ServiceManager helper code for managing events for the HLE services.

**Comment by Morph1984** [comment]
In addition to the comments above - please fixup the `Keep share memory as close` commit with the `Fix share memory object leak` commit

**Comment by Morph1984** [comment]
I've replaced all `KAutoObject::Create(event)` with `KernelHelpers::ServiceContext::CreateEvent` in #7103

**Comment by vonchenplus** [comment]
OK, I revert the relevant changes.

**Review by bunnei — APPROVED** [review]

**Review by ameerj — APPROVED** [review]

**Pull request merged**